File: packages/python/src/Seguranca/Ataques/AtaqueSAC.py
"""SAC (Strict Avalanche Criterion): para CADA bit de entrada (IV), ao virar
esse bit, CADA bit de saída deve mudar com probabilidade ~0.5. Mede o pior
bit de saída; se algum fica longe de 50%, a difusão tem ponto cego.
"""

from ..ResultadoAtaque import ResultadoAtaque
from ..Util import random_bytes


class AtaqueSAC:
    def __init__(self, tamanho_bloco: int = 32, amostras: int = 40, tolerancia: float = 0.05):
        self.tamanho_bloco = tamanho_bloco
        self.amostras = amostras
        self.tolerancia = tolerancia

    def nome(self) -> str:
        return "SAC (Strict Avalanche Criterion)"

    def executar(self, alvo) -> ResultadoAtaque:
        chave = alvo.chave_de_teste().encode()
        total_bits = self.tamanho_bloco * 8
        flips = [0] * total_bits
        tamanho_iv = alvo.tamanho_iv()
        total = 0

        for pos in range(tamanho_iv):
            for bit in range(8):
                for _ in range(self.amostras):
                    iv1 = random_bytes(tamanho_iv)
                    iv2 = bytearray(iv1)
                    iv2[pos] ^= 1 << bit

                    ks1 = alvo.gerar_keystream_bruto(chave, iv1, "enc", self.tamanho_bloco)
                    ks2 = alvo.gerar_keystream_bruto(chave, bytes(iv2), "enc", self.tamanho_bloco)

                    for j in range(total_bits):
                        b1 = (ks1[j >> 3] >> (7 - (j & 7))) & 1
                        b2 = (ks2[j >> 3] >> (7 - (j & 7))) & 1
                        if b1 != b2:
                            flips[j] += 1
                    total += 1

        pior = -1
        pior_desvio = 0.0
        pior_p = 0.5
        for j in range(total_bits):
            p = flips[j] / total
            desvio = abs(p - 0.5)
            if desvio > pior_desvio:
                pior_desvio = desvio
                pior = j
                pior_p = p

        vulneravel = pior_desvio > self.tolerancia
        detalhes = (
            f"pior bit de saída={pior}: p={pior_p * 100:.2f}% "
            f"(esperado 50%, tolerância ±{self.tolerancia * 100:.1f}%); "
            f"{total} amostras por bit de entrada"
        )

        return ResultadoAtaque(
            self.nome(),
            vulneravel,
            "alta" if vulneravel else "info",
            detalhes,
            {"pior": pior, "pior_p": pior_p, "desvio": pior_desvio},
        )

File: packages/dart/lib/Seguranca/Ataques/AtaqueSAC.dart
import 'dart:typed_data';

import 'package:criptografia/Seguranca/AlvoCriptografico.dart';
import 'package:criptografia/Seguranca/Ataque.dart';
import 'package:criptografia/Seguranca/ResultadoAtaque.dart';
import 'package:criptografia/Seguranca/Util.dart';

/// SAC (Strict Avalanche Criterion): para CADA bit de entrada (IV), ao virar
/// esse bit, CADA bit de saída deve mudar com probabilidade ~0.5. Mede o pior
/// bit de saída; se algum fica longe de 50%, a difusão tem ponto cego.
class AtaqueSAC extends Ataque {
  final int _tamanhoBloco;
  final int _amostras;
  final double _tolerancia;

  AtaqueSAC({
    int tamanhoBloco = 32,
    int amostras = 40,
    double tolerancia = 0.05,
  }) : _tamanhoBloco = tamanhoBloco,
       _amostras = amostras,
       _tolerancia = tolerancia;

  @override
  String nome() => 'SAC (Strict Avalanche Criterion)';

  @override
  ResultadoAtaque executar(AlvoCriptografico alvo) {
    final chave = toBytes(alvo.chaveDeTeste());
    final totalBits = _tamanhoBloco * 8;
    final flips = List<int>.filled(totalBits, 0);
    final tamanhoIv = alvo.tamanhoIv();
    var total = 0;

    for (var pos = 0; pos < tamanhoIv; pos++) {
      for (var bit = 0; bit < 8; bit++) {
        for (var s = 0; s < _amostras; s++) {
          final iv1 = randomBytes(tamanhoIv);
          final iv2 = Uint8List.fromList(iv1);
          iv2[pos] ^= 1 << bit;

          final ks1 = alvo.gerarKeystreamBruto(
            chave,
            iv1,
            'enc',
            _tamanhoBloco,
          );
          final ks2 = alvo.gerarKeystreamBruto(
            chave,
            iv2,
            'enc',
            _tamanhoBloco,
          );

          for (var j = 0; j < totalBits; j++) {
            final b1 = (ks1[j >> 3] >> (7 - (j & 7))) & 1;
            final b2 = (ks2[j >> 3] >> (7 - (j & 7))) & 1;
            if (b1 != b2) {
              flips[j]++;
            }
          }
          total++;
        }
      }
    }

    var pior = -1;
    var piorDesvio = 0.0;
    var piorP = 0.5;
    for (var j = 0; j < totalBits; j++) {
      final p = flips[j] / total;
      final desvio = (p - 0.5).abs();
      if (desvio > piorDesvio) {
        piorDesvio = desvio;
        pior = j;
        piorP = p;
      }
    }

    final vulneravel = piorDesvio > _tolerancia;
    final detalhes =
        'pior bit de saída=$pior: p=${(piorP * 100).toStringAsFixed(2)}% '
        '(esperado 50%, tolerância ±'
        '${(_tolerancia * 100).toStringAsFixed(1)}%); '
        '$total amostras por bit de entrada';

    return ResultadoAtaque(
      nome(),
      vulneravel,
      vulneravel ? Severidade.alta : Severidade.info,
      detalhes,
      {'pior': pior, 'pior_p': piorP, 'desvio': piorDesvio},
    );
  }
}
